### PR TITLE
[Auto Parallel] Compatible new comm library upgrade for XPUs.

### PR DESCRIPTION
**Pull request by ZibinGuo**
### PR Category
Communication Library

### PR Types
New features

### Description
根据[#56604](https://github.com/PaddlePaddle/Paddle/pull/56604) pr，在xpu上适配新版静态图分布式通信库

### CODE REVIEW

File: python/paddle/distributed/launch/context/device.py [hunk 1]
```diff
@@ -102,7 +102,10 @@ def parse_device(self):
...
                 visible_devices = os.getenv(visible_devices_str)
         elif 'CUDA_VISIBLE_DEVICES' in os.environ:
-            dev._dtype = DeviceType.GPU
+            if core.is_compiled_with_xpu():
+                dev._dtype = DeviceType.XPU
+            else:
+                dev._dtype = DeviceType.GPU
             visible_devices = os.getenv("CUDA_VISIBLE_DEVICES")
         elif 'XPU_VISIBLE_DEVICES' in os.environ:
             dev._dtype = DeviceType.XPU
```
Comment (XiaociZhang) [inline python/paddle/distributed/launch/context/device.py]: 可能要再加一个XPULINK_VISIBLE_DEVICES

**Review by RuohengMa — APPROVED** [review]
LGTM

File: python/paddle/distributed/launch/context/device.py [hunk 2]
```diff
@@ -101,12 +101,18 @@ def parse_device(self):
...
-            visible_devices = os.getenv("CUDA_VISIBLE_DEVICES")
+        elif 'XPULINK_VISIBLE_DEVICES' in os.environ:
+            dev._dtype = DeviceType.XPU
+            visible_devices = os.getenv("XPULINK_VISIBLE_DEVICES")
         elif 'XPU_VISIBLE_DEVICES' in os.environ:
             dev._dtype = DeviceType.XPU
             visible_devices = os.getenv("XPU_VISIBLE_DEVICES")
+        elif 'CUDA_VISIBLE_DEVICES' in os.environ:
+            if core.is_compiled_with_xpu():
+                dev._dtype = DeviceType.XPU
```
Comment (houj04) [inline python/paddle/distributed/launch/context/device.py:112]: 这里后续考虑加个注释？因为不知道背景的人可能会觉得疑惑，为什么在XPU下面会刷CUDA的环境变量。

**Review by houj04 — APPROVED** [review]

File: python/paddle/distributed/launch/context/device.py (hunk 2 above)
Comment (ZibinGuo) [reply python/paddle/distributed/launch/context/device.py:112]: 好的，下个pr里面加上

File: python/paddle/distributed/launch/context/device.py (hunk 1 above)
Comment (dynamicheart) [reply python/paddle/distributed/launch/context/device.py]: @XiaociZhang 

PP8时，需要设置：
```bash
export CUDA_DEVICE_ORDER=OAM_ID
export XPULINK_VISIBLE_DEVICES=2,3,0,1,5,4,7,6
```

然而，会导致rank0对应dev2，进而导致通信库无法正常工作，由 @Thunderbrook  进行问题的排查，以下代码进行相关说明：

这里传入的devices是模型启动脚本里面的--xpus，为0,1,2,3,4,5,6,7：
- https://github.com/PaddlePaddle/Paddle/blob/827f3622392d9914891be059da1fe69fb4668f6c/python/paddle/distributed/launch/controllers/collective.py#L114
- https://github.com/PaddlePaddle/Paddle/blob/827f362/python/paddle/distributed/launch/context/args_envs.py#L138

这里的device._labels是从XPULINK_VISIBLE_DEVICES解析，为2,3,0,1,5,4,7,6，get_selected_devices也为2,3,0,1,5,4,7,6，因此rank0为dev2
- https://github.com/PaddlePaddle/Paddle/blob/827f362/python/paddle/distributed/launch/context/node.py#L27
- https://github.com/PaddlePaddle/Paddle/blob/827f362/python/paddle/distributed/launch/context/device.py#L106
- https://github.com/PaddlePaddle/Paddle/blob/827f362/python/paddle/distributed/launch/context/device.py#L85


解决方案：

方案1：设置`export XPULINK_VISIBLE_DEVICES=2,3,0,1,5,4,7,6`的同时，需要设置`--xpus "2,3,0,1,5,4,7,6"`，这样使得rank0，仍然为dev0

方案2(推荐方案)：训练参数去掉`--xpus`


**综上，机内PP8需要**：
1. 设置环境变量
```bash
export CUDA_DEVICE_ORDER=OAM_ID
export XPULINK_VISIBLE_DEVICES=2,3,0,1,5,4,7,6
```

2. 训练参数去掉`--xpus`

**Comment by ZibinGuo** [comment]
您的邮件已收到，我将尽快回信。谢谢您！Your email has been received and I will reply as soon as possible. Thank you！

File: python/paddle/distributed/launch/context/device.py (hunk 1 above)
Comment (XiaociZhang) [reply python/paddle/distributed/launch/context/device.py]: 方案2 is preferred

**Pull request merged**